vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSPackageRegistry.java

```diff
@@ -0,0 +1,570 @@
+import org.apache.jackrabbit.vault.fs.io.ImportOptions;
+import org.apache.jackrabbit.vault.fs.io.MemoryArchive;
+import org.apache.jackrabbit.vault.packaging.Dependency;
+import org.apache.jackrabbit.vault.packaging.NoSuchPackageException;
+import org.apache.jackrabbit.vault.packaging.PackageException;
+import org.apache.jackrabbit.vault.packaging.PackageExistsException;
+import org.apache.jackrabbit.vault.packaging.PackageId;
+import org.apache.jackrabbit.vault.packaging.VaultPackage;
+import org.apache.jackrabbit.vault.packaging.events.PackageEvent;
+import org.apache.jackrabbit.vault.packaging.events.impl.PackageEventDispatcher;
+import org.apache.jackrabbit.vault.packaging.impl.PackagePropertiesImpl;
+import org.apache.jackrabbit.vault.packaging.impl.ZipVaultPackage;
+import org.apache.jackrabbit.vault.packaging.registry.DependencyReport;
+import org.apache.jackrabbit.vault.packaging.registry.ExecutionPlanBuilder;
+import org.apache.jackrabbit.vault.packaging.registry.PackageRegistry;
+import org.apache.jackrabbit.vault.packaging.registry.RegisteredPackage;
+import org.apache.jackrabbit.vault.util.InputStreamPump;
+import org.apache.jackrabbit.vault.util.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@code JcrPackagePersistence}...
+ */
+public class FSPackageRegistry implements PackageRegistry, InternalPackageRegistry {
+
+    /**
+     * default logger
+     */
+    private static final Logger log = LoggerFactory.getLogger(FSPackageRegistry.class);
+
+
+    /**
+     * default root path for packages
+     */
+    public static final String DEFAULT_PACKAGE_ROOT_PATH = "/etc/packages";
+
+    /**
+     * default root path prefix for packages
+     */
+    public static final String DEFAULT_PACKAGE_ROOT_PATH_PREFIX = DEFAULT_PACKAGE_ROOT_PATH + "/";
+
+    private final String[] SUFFIXES = {"zip"};
+    
+
+    @Nullable
+    private PackageEventDispatcher dispatcher;
+
+    private static File homeDir;
+    private static File statusFile;
+    
+    private volatile Map<String, InstallState> statusList = Collections.emptyMap();
+
+    public File getHomeDir() {
+        return homeDir;
+    }
+    
+    
+
+    /**
+     * Creates a new FSPackageRegistry based on the given session.
+     * 
+     * @param session
+     *            the JCR session that is used to access the repository.
+     * @param roots
+     *            the root paths to store the packages.
+     * @throws IOException
+     */
+    @SuppressWarnings("unchecked")
+    public FSPackageRegistry(@Nonnull File homeDir) throws IOException {
+        FSPackageRegistry.homeDir = homeDir;
+        if (homeDir.exists()) {
+            FSPackageRegistry.statusFile = new File(homeDir, "registrystatus.ser");
+            if (statusFile.exists()) {
+                FileInputStream fis = null;
+                ObjectInputStream ois = null;
+                try {
+                    fis = new FileInputStream("hashmap.ser");
```

Comment:
instead of 1 serialized hashmap, I think it's more robust to have a status file (properties or json) for each package. the file traversal overhead is not huge if it only happens once when a registry is created (assuming it can be cached as a service, for example). having individual files reduces concurrency issues.